chcemvediet/apps/inforequests/tests/test_cron.py

```diff
@@ -57,7 +55,10 @@ def _call_cron_job(self):
 
 
     def test_times_job_is_run_at(self):
-        self.assert_times_job_is_run_at(u'chcemvediet.apps.inforequests.cron.undecided_email_reminder')
+        self.assert_times_job_is_run_at(
+                u'chcemvediet.apps.inforequests.cron.undecided_email_reminder',
+                settings.CRON_USER_INTERACTION_TIMES
```

Comment:
Nie je moc dobre robit testy oproti dynamickej premennej, ktora je pouzivaja aj priamo v implementacii. Lepsie by bolo `settings.CRON_USER_INTERACTION_TIMES` najprv overridnut na nieco fixne a potom mat v teste fixne casy s fixnymi rucne urcenymi hodnotami, ci v danom case sa ma cron spustit, ci nie.

Tento test nema testovat, ci test sa spusti v konkretnom case. Ale ci implementacia testu spravne pouziva `settings.CRON_USER_INTERACTION_TIMES`.

Nieco cca taketo:

```
with self.settings(CRON_USER_INTERACTION_TIMES=[...]):
  self.assert_times_job_is_run_at('chcemvediet.apps.inforequests.cron.undecided_email_reminder', (
    (u'07:10', False), (u'07:50', False), ...
    (u'09:10', True),  (u'09:50', False), ...
    ...
  ))
```

Pricom mnozstvo vzoriek `(u'07:10', False), (u'07:50', False), ...` nemusi byt az take velke ako bolo doteraz. Dolezite je hlavne aby boli otestovane casy v okoli hranicnych hodnot.

Reply:
@martinmacko47 Nebude to takto fungovat, pretoze `def undecided_email_reminder():` spolu s `@cron_job(run_at_times=settings.CRON_USER_INTERACTION_TIMES)` sa skompiluje skor ako testy. V cron funkcii budu inicializovane povodne hodnoty, override settingov v teste nema ziaden vplyv.

Reply:
@viliambalaz To sa da vyriesit [prekompilovanim modulu] `chcemvediet.apps.inforequests.cron` po zmene settingov:

[prekompilovanim modulu]: https://docs.python.org/2.7/library/functions.html#reload

```
from .. import cron

def test_xxx(...):
  with self.settings(CRON_USER_INTERACTION_TIMES=[...]):
    reload(cron)
    self.assert_times_job_is_run_at(...)
  reload(cron)
```

Pozor, ked sa na konci overridu zmenia settingy naspat na povodne, tak `reload(cron)` treba zavolat znovu. Nech sa modul skompiluje naspat s povodnymi settingami.

Mozno by sme mohli na to spravit vlastny kontext manager, ktory dostane ako parameter iny context manager a zoznam modulov, ktore ma reloadnut. Teda nieco ako:

```
with reload_for_context(self.settings(...), cron):
    self.assert_times_job_is_run_at(...) 
```

Implementacia by bola cca nieco ako, len treba do doladit:

```
@contextlib.contextmanager
def reload_for_context(manager, module):
  with manager as context:
    reload(module)
    yield context
  reload(module)
```

Pripadne by sme mohli dokonca skusit namiesto modulu posielat ako parameter iba cestu k modulu. Aby ten modul nebolo treba pri pouziti kontext manageru vzdy importovat. Namiesto toho by si modul vzdy najprv importol kontext manager pomocou [`importlib.import_module`].

[`importlib.import_module`]: https://docs.python.org/2.7/library/importlib.html#importlib.import_module

Takze by sme to mohli pouzit ako:

```
with reload_for_context(self.settings(...), u'chcemvediet.apps.inforequests.cron'):
    self.assert_times_job_is_run_at(...) 
```

Reply:
Skusal som to aj pomocou `del sys.modules['chcemvediet.apps.inforequests.cron']`. Samostatne pri spusani jednotlivych testov to fungovalo, ale ked som spustil vsetky naraz, tak to nefungovalo. Bol problem ze `'chcemvediet.apps.inforequests.cron'` nebol spravne importnuty. Preto som pouzil riesenie s `reload`.

Reply:
:ok: